src/api/index.ts

```diff
@@ -0,0 +1,191 @@
+}
+
+/**
+ * Definition for the arguments you need to call a particular API endpoint
+ */
+interface ApiCallDef<E extends ApiEndpoint, Params, Body> {
+  endpoint: E;
+  params: Params;
+  body: Body;
+}
+type TrustMemberApiCall = ApiCallDef<
+  ApiEndpoint.TRUST_MEMBER,
+  { uid: Uid },
+  void
+>;
+type GetOperationsApiCall = ApiCallDef<ApiEndpoint.GET_OPERATIONS, void, void>;
+/**
+ * All API calls you can make, and the arguments you need to call them.
+ */
+export type ApiCall = TrustMemberApiCall | GetOperationsApiCall;
+
+/**
+ * Definition of how to use an API endpoint, i.e. what you have to provide to
+ * call it, and what it will return to you.
+ */
+interface ApiEndpointDef<Call extends ApiCall, Resp extends ApiResponse> {
+  call: Call;
+  response: Resp;
+}
+
+export type TrustMemberApiEndpoint = ApiEndpointDef<
+  TrustMemberApiCall,
+  OperationApiResponse
+>;
+
+export type GetOperationsApiEndpoint = ApiEndpointDef<
+  GetOperationsApiCall,
+  OperationsApiResponse
+>;
+type ApiDefinition = TrustMemberApiEndpoint | GetOperationsApiEndpoint;
+
+/* =================================
+ * Resolving API endpoint locations
+ * =================================
+ */
+enum HttpVerb {
+  GET = "GET",
+  POST = "POST",
+  PUT = "PUT",
+  DELETE = "DELETE",
+  PATCH = "PATCH"
+}
+
+/**
+ * The location of an API endpoint. Uri may contain wildcards that must be
+ * resolved, and only represents a path without a domain to send the request to.
+ */
+interface ApiEndpointSpec {
+  uri: string;
+  method: HttpVerb;
+}
+
+/**
+ * Mapping from API endpoints to their corresponding unresolved locations
+ */
+const apiEndpointSpecs: { [key in ApiEndpoint]: ApiEndpointSpec } = {
+  [ApiEndpoint.TRUST_MEMBER]: {
+    uri: "members/:uid/trust",
+    method: HttpVerb.POST
+  },
+  [ApiEndpoint.GET_OPERATIONS]: {
+    uri: "operations",
+    method: HttpVerb.GET
+  }
+};
+
+/**
+ * The location of an API endpoint, with a full url to query and parameters
+ * resolved.
+ */
+interface ResolvedApiEndpointSpec {
+  url: string;
+  method: HttpVerb;
+}
+
+/**
+ * Determines the URL and HTTP method for an API call.
+ */
+export function resolveApiEndpoint<Def extends ApiDefinition>(
+  apiCall: Def["call"]
+): ResolvedApiEndpointSpec {
+  const { uri, method } = apiEndpointSpecs[apiCall.endpoint];
+
+  const params = apiCall.params;
+  if (!params) {
+    return { url: `${API_BASE}${uri}`, method };
+  }
+
+  const wildcards = uri.split("/").filter(part => part.charAt(0) === ":");
+
+  const resolvedUri = wildcards.reduce((memo, wildcard) => {
+    const paramName = wildcard.slice(1);
+    if (!(paramName in params)) {
+      throw new InvalidApiRequestError(apiCall);
+    }
+    const paramValue: string = (apiCall.params as any)[paramName];
+    // TODO: [#security] is there any sanitization we need to do here?
+    return memo.replace(wildcard, paramValue);
+  }, uri);
+
+  return { url: `${API_BASE}${resolvedUri}`, method };
+}
+
+/**
+ * Call an API endpoint.
```

Comment:
Note that this parses the response as JSON and will throw an error if the response is not valid JSON.